### PR TITLE
testing/synergy: new aport

### PR DESCRIPTION
**Pull request by dlintw**
for build synergy, it is required to build new 'gmock' & modify 'gtest' packages.

### CODE REVIEW

**Comment by dlintw** [comment]
The automatic CI can not pass because, synergy required gtest & gmock build and install. The CI process only build without installation.

File: testing/gtest/APKBUILD [hunk 1]
```diff
@@ -0,0 +1,55 @@
+# Maintainer: Francesco Colista <fcolista@alpinelinux.org>
```
Comment (kaniini) [inline testing/gtest/APKBUILD]: Why are we introducing a new gtest package to testing?

File: testing/synergy/APKBUILD [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+system-gtest.patch
+missing_sentinel.patch
+"
+#builddir="$srcdir/"
+builddir="$srcdir/synergy-${pkgver}-stable"
+
+prepare() {
+  cd "$builddir"
+  mkdir -p build
+  patch -p1 < "${srcdir}/system-gtest.patch"
```
Comment (kaniini) [inline testing/synergy/APKBUILD]: use `default_prepare || return 1` to apply the patches automatically.

File: testing/synergy/APKBUILD [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  patch -p0 < "${srcdir}/missing_sentinel.patch"
```
Comment (kaniini) [inline testing/synergy/APKBUILD]: all patches should apply with `-p1`

File: testing/synergy/APKBUILD [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  msg2 "Building GUI..."
+  cd ..
+  cd src/gui
+  qmake-qt5 || return 1
+  make || return 1
+}
+check() {
+  cd "$builddir"
+  pwd
+  ./bin/unittests || return
```
Comment (kaniini) [inline testing/synergy/APKBUILD]: `./bin/unittests || return 1` so that we signal failure if the unit tests fail

File: testing/synergy/APKBUILD [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+  install -Dm 755 bin/synergy bin/synergyc bin/synergyd bin/synergys bin/syntool bin/usynergy -t "${pkgdir}/usr/bin" || return
+
+  # install config
+  install -Dm 644 doc/${pkgname}.conf* -t "${pkgdir}/etc" || return
+
+  # install manfiles
+  install -Dm 644 "doc/${pkgname}c.man" "${pkgdir}/usr/share/man/man1/${pkgname}c.1" || return
+  install -Dm 644 "doc/${pkgname}s.man" "${pkgdir}/usr/share/man/man1/${pkgname}s.1" || return
+
+  # install systemd service and socket
```
Comment (kaniini) [inline testing/synergy/APKBUILD]: no need to retain this section, we don't use systemd

File: testing/synergy/missing_sentinel.patch [hunk 6]
```diff
@@ -0,0 +1,78 @@
+--- src/lib/platform/XWindowsScreen.cpp.orig	2017-02-27 21:26:49.297547435 +0800
```
Comment (kaniini) [inline testing/synergy/missing_sentinel.patch]: Change this to a `-p1` style patch, by changing it to `--- a/src/lib/platform/XWindowsScreen.cpp` and `+++ b/src/lib/platform/XWindowsScreen.cpp` in the patch header.

File: testing/gmock/APKBUILD [hunk 7]
```diff
@@ -0,0 +1,49 @@
+# Contributor: Daniel YC Lin <dlin.tw@gmail.com>
+# Maintainer: Daniel YC Lin <dlin.tw@gmail.com>
+pkgname=gmock
+pkgver=1.8.0
+pkgrel=1
```
Comment (kaniini) [inline testing/gmock/APKBUILD]: pkgrel should start at 0

File: testing/gmock/APKBUILD [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+source="https://github.com/google/googletest/archive/release-$pkgver.tar.gz
+  gmock-install.patch
+  gmock.pc.in
+"
+builddir="$srcdir"/googletest-release-${pkgver}
+
+prepare() {
+  cd "$builddir"
+  find -name '*.py' -exec sed -i 's|env python|env python2|g' {} \;
+  patch -p1 < "${srcdir}/gmock-install.patch"
```
Comment (kaniini) [inline testing/gmock/APKBUILD]: use `default_prepare || return 1` to apply the patch

**Comment by kaniini** [comment]
Please make the requested changes and flatten down the commits into a single commit.

File: testing/gtest/APKBUILD (hunk 1 above)
Comment (dlintw) [inline testing/gtest/APKBUILD]: The original gtest missing the /usr/src, that is required for gtest's user.

File: testing/synergy/APKBUILD (hunk 4 above)
Comment (dlintw) [inline testing/synergy/APKBUILD]: This is a short form of return 1, try the following script, you will got it.
```
#!/bin/sh
asfalse() {
  false || return
}
astrue() {
  true && return
}
astrue
echo "$?"
asfalse
echo "$?"

```

File: testing/gtest/APKBUILD (hunk 1 above)
Comment (dlintw) [inline testing/gtest/APKBUILD]: In fact, in Arch Linux, the gtest & gmock is combined into one PKGBUILD. But, I don't know how to do it proper in Alpine Linux.

**Comment by kaniini** [comment]
the testing/gtest changes should just be merged into main/gtest.

**Comment by jirutka** [comment]
I guess that @kaniini overlooked that this PR adds two packages and modifies one package, so it should not be flattened into a single commit. It should be a single commit per package, with commit message in format `<repo>/<pkgname>: …`.

File: testing/gmock/APKBUILD [hunk 9]
```diff
@@ -0,0 +1,51 @@
+# Contributor: Daniel YC Lin <dlin.tw@gmail.com>
+# Maintainer: Daniel YC Lin <dlin.tw@gmail.com>
+pkgname=gmock
+pkgver=1.8.0
+pkgrel=0
+pkgdesc="Google Mock - A library for writing and using C++ mock classes"
+url="https://github.com/google/googletest"
+arch="all"
+license="BSD"
+depends="libgcc bash python2 gtest"
```
Comment (jirutka) [inline testing/gmock/APKBUILD]: Does it really need bash? How many scripts are not POSIX-sh compatible?

Comment (jirutka) [inline testing/gmock/APKBUILD]: Dependency on libgcc is auto-discovered via dynamic links (`so:libgcc_s.so.1`), so it’s not needed (should not be) to specify it here.

File: testing/gmock/APKBUILD [hunk 10]
```diff
@@ -0,0 +1,51 @@
+# Contributor: Daniel YC Lin <dlin.tw@gmail.com>
```
Comment (jirutka) [inline testing/gmock/APKBUILD]: Please use only tabs for indentation (and alignment) in APKBUILDs and other shell scripts, not spaces.

File: testing/gmock/APKBUILD [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+  find -name '*.py' -exec sed -i 's|env python|env python2|g' {} \;
+  default_prepare || return
+  cd googlemock
+  cp "${srcdir}/gmock.pc.in" .
+  autoreconf -fvi || return
+}
+
+build() {
+  cd "$builddir"
+  cd googlemock
```
Comment (jirutka) [inline testing/gmock/APKBUILD]: Why not just `cd "$builddir"/googlemock`…?

File: testing/synergy/APKBUILD [hunk 12]
```diff
@@ -0,0 +1,64 @@
+# Contributor: Daniel YC Lin <dlin.tw@gmail.com>
```
Comment (jirutka) [inline testing/synergy/APKBUILD]: Please use only tabs for indentation (and alignment) in APKBUILDs and other shell scripts, not spaces.

File: testing/gmock/APKBUILD [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+  ./configure --prefix=/usr --with-gtest --enable-external-gtest || return
+  make || return
+}
+
+package() {
+  cd "$builddir"
+  cd googlemock
+  make --trace -w -j1 DESTDIR="${pkgdir}" install || return
```
Comment (jirutka) [inline testing/gmock/APKBUILD]: This installs even files that should be moved into -doc subpackage:

/usr/share/gmock/generator/LICENSE
/usr/share/gmock/generator/README
/usr/share/gmock/generator/README.cppclean

**Comment by jirutka** [comment]
I wonder why buildrepo skipped main/gtest, this is probably some bug. Therefore Travis failed to build testing/synergy.

**Comment by dlintw** [comment]
I'm wonder if I separate the gtest,gmock,synergy into three pull request, it should be failed in CI.  I just moving the testing/gtest to main/gtest, But it can not build synergy, because main/gtest can not be update correctly.

File: testing/gmock/APKBUILD (hunk 10 above)
Comment (dlintw) [inline testing/gmock/APKBUILD]: done

File: testing/gmock/APKBUILD (hunk 9 above)
Comment (dlintw) [inline testing/gmock/APKBUILD]: I'm not sure, I've removed bah, Let CI to try it.

Comment (dlintw) [inline testing/gmock/APKBUILD]: done

File: testing/gmock/APKBUILD (hunk 11 above)
Comment (dlintw) [inline testing/gmock/APKBUILD]: done

File: testing/gmock/APKBUILD (hunk 13 above)
Comment (dlintw) [inline testing/gmock/APKBUILD]: done

File: testing/synergy/APKBUILD (hunk 12 above)
Comment (dlintw) [inline testing/synergy/APKBUILD]: done

**Comment by jirutka** [comment]
I haven't said anything about separating it into three pull requests, just commits...

File: testing/gmock/APKBUILD (hunk 8 above)
Comment (dlintw) [inline testing/gmock/APKBUILD]: done

File: testing/gmock/APKBUILD [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+	  -t "${pkgdir}/usr/share/doc/${pkgname}" || return
+	install -Dm 755 scripts/gmock-config -t "${pkgdir}/usr/bin" || return
+	install -Dm 644 include/gmock/internal/custom/* \
+	  -t "${pkgdir}/usr/include/gmock/internal/custom" || return
+	mkdir "${pkgdir}/usr/share/doc/${pkgname}/generator"
+	mv "${pkgdir}/usr/share/gmock//generator/LICENSE" \
+	  "${pkgdir}/usr/include/gmock/internal/custom" || return
+	mv "${pkgdir}/usr/share/gmock//generator/README" \
+	  "${pkgdir}/usr/include/gmock/internal/custom" || return
+	mv "${pkgdir}/usr/share/gmock//generator/README.cppclean \
```
Comment (jirutka) [inline testing/gmock/APKBUILD]: There’s closing quote missing. I wonder why Travis skipped this abuild. o.O

**Comment by dlintw** [comment]
forgive it

**Pull request closed without merging**